From ec3c8ab0bcd8c193abccbe9f6512870e7ca0bf82 Mon Sep 17 00:00:00 2001
From: Jiangli Yuan <jlyuan@motorola.com>
Date: Wed, 4 Mar 2015 18:24:08 -0800
Subject: [PATCH 766/959] IKSWL-41240: sps: make sure pipe active mask match
 with pipe list status

 The pipe may get removed but pipe active mask is not update yet.
 Add protection to make sure those two are matched.

Change-Id: If0e2f21be278151d69124425c8aa6506634669e2
Signed-off-by: Jiangli Yuan <jlyuan@motorola.com>
Reviewed-on: http://gerrit.mot.com/721532
SLTApproved: Slta Waiver <sltawvr@motorola.com>
Tested-by: Jira Key <jirakey@motorola.com>
Reviewed-by: Zhi-Ming Yuan <a14194@motorola.com>
Reviewed-by: Sheng-Zhe Zhao <a18689@motorola.com>
Submit-Approved: Jira Key <jirakey@motorola.com>
Reviewed-on: http://gerrit.mot.com/830284
SME-Granted: SME Approvals Granted
Reviewed-by: Igor Kovalenko <igork@motorola.com>
---
 drivers/platform/msm/sps/sps_bam.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/platform/msm/sps/sps_bam.c b/drivers/platform/msm/sps/sps_bam.c
index bc039f98f4b..ec59fd449e4 100644
--- a/drivers/platform/msm/sps/sps_bam.c
+++ b/drivers/platform/msm/sps/sps_bam.c
@@ -1030,12 +1030,12 @@ int sps_bam_pipe_connect(struct sps_pipe *bam_pipe,
 	}
 
 	/* Indicate initialization is complete */
+	spin_lock_irqsave(&dev->isr_lock, flags);
 	dev->pipes[pipe_index] = bam_pipe;
 	dev->pipe_active_mask |= 1UL << pipe_index;
-	spin_lock_irqsave(&dev->isr_lock, flags);
 	list_add_tail(&bam_pipe->list, &dev->pipes_q);
-	spin_unlock_irqrestore(&dev->isr_lock, flags);
 	bam_pipe->state |= BAM_STATE_INIT;
+	spin_unlock_irqrestore(&dev->isr_lock, flags);
 	result = 0;
 exit_err:
 	if (result) {
@@ -1076,8 +1076,8 @@ int sps_bam_pipe_disconnect(struct sps_bam *dev, u32 pipe_index)
 			unsigned long flags = 0;
 			spin_lock_irqsave(&dev->isr_lock, flags);
 			list_del(&pipe->list);
-			spin_unlock_irqrestore(&dev->isr_lock, flags);
 			dev->pipe_active_mask &= ~(1UL << pipe_index);
+			spin_unlock_irqrestore(&dev->isr_lock, flags);
 		}
 		dev->pipe_remote_mask &= ~(1UL << pipe_index);
 		if (pipe->connect.options & SPS_O_NO_DISABLE)
@@ -2326,10 +2326,10 @@ int sps_bam_set_satellite(struct sps_bam *dev, u32 pipe_index)
 	/* Indicate satellite control */
 	spin_lock_irqsave(&dev->isr_lock, flags);
 	list_del(&pipe->list);
-	spin_unlock_irqrestore(&dev->isr_lock, flags);
 	dev->pipe_active_mask &= ~(1UL << pipe_index);
 	dev->pipe_remote_mask |= pipe->pipe_index_mask;
 	pipe->state |= BAM_STATE_REMOTE;
+	spin_unlock_irqrestore(&dev->isr_lock, flags);
 
 	return 0;
 }
-- 
2.11.0

